spark/src/main/java/org/apache/iceberg/spark/source/StreamingOffset.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.io.StringWriter;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+
+/**
+ * An implementation of Spark Structured Streaming Offset, to track the current processed
+ * files of Iceberg table. This StreamingOffset is combined by:
+ *
+ * version: used to validate the version number of StreamingOffset when deserializing from json
+ * string.
+ * snapshot_id: used to record the current processed snapshot of Iceberg table.
+ * index: used to record the index of scan files in this snapshot.
+ * is_starting_snapshot: used to identify if this snapshot is a starting snapshot. For starting snapshot
+ * we should scan all the files, not just incrementally add files.
+ * last_index_of_snapshot: used to denote whether the current index is the last file of this snapshot,
+ * to avoid revist the processed snapshot.
+ */
+class StreamingOffset extends Offset {
+  static final StreamingOffset START_OFFSET = new StreamingOffset(-1L, -1, false, true);
+
+  private static final int CURR_VERSION = 1;
+  private static final String VERSION = "version";
+  private static final String SNAPSHOT_ID = "snapshot_id";
+  private static final String INDEX = "index";
+  private static final String IS_STARTING = "is_starting_snapshot";
+  private static final String LAST_INDEX_OF_SNAPSHOT = "last_index_of_snapshot";
+
+  private final long snapshotId;
+  private final int index;
+  private final boolean isStarting;
```

Comment:
I think rather than "isStarting" this should be something like scanAllFiles. Since the condition in which we scan allFiles is less important than the fact that we are scanning all files.

Reply:
Sure, `scanAllFiles` seems more straightforward.